# **1. Import Library**

Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [134]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler, MinMaxScaler
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

# **2. Memuat Dataset dari Hasil Clustering**

Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [135]:
file_id = '16alhprVK9jQGGXjUp6qbiKiYFUlwr9Hx'
download_url = f'https://drive.google.com/uc?id={file_id}'

# Baca file CSV dari URL
data = pd.read_csv(download_url)

# Tampilkan DataFrame untuk memastikan telah dibaca dengan benar
data.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

    Reviews  Size    Installs  Type  Price Content Rating  \
0     159.0   19M     10000.0  Free    0.0       Everyone   
1     967.0   14M    500000.0  Free    0.0       Everyone   
2   87510.0  8.7M   5000000.0  Free    0.0       Everyone   
3  215644.0   25M  50000000.0  Free    0.0           Teen   
4     967.0  2.8M    100000.0  Free    0.0       Everyone   

                      Genres Last Updated         Current Ver   Android Ver  \
0               Art & Design   2018-01-07               1.0.0  4.0.3 and up   
1  Art & Design;Pretend Play   2018-01-15               2.0.0  4.0.3 and up   
2               Art & Design   2018-08-01               1.2.4  4.0.3 and up   
3               Art & Design   2018-06-08  Varies with device    4.2 and up   
4    Art & Design;Creativity   2018-06-20                 1.1    4.4 and up   

   Cluster  
0        0  
1        1  
2        0  
3        0  
4        0

In [136]:
# Tampilkan informasi umum tentang dataset
print("\nInformasi dataset:")
data.info()


Informasi dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8886 entries, 0 to 8885
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8886 non-null   object 
 1   Category        8886 non-null   object 
 2   Rating          8886 non-null   float64
 3   Reviews         8886 non-null   float64
 4   Size            8886 non-null   object 
 5   Installs        8886 non-null   float64
 6   Type            8886 non-null   object 
 7   Price           8886 non-null   float64
 8   Content Rating  8886 non-null   object 
 9   Genres          8886 non-null   object 
 10  Last Updated    8886 non-null   object 
 11  Current Ver     8886 non-null   object 
 12  Android Ver     8886 non-null   object 
 13  Cluster         8886 non-null   int64  
dtypes: float64(4), int64(1), object(9)
memory usage: 972.0+ KB


In [137]:
# Cek missing values
print("\nMissing values per fitur:")
print(data.isnull().sum())


Missing values per fitur:
App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
Cluster           0
dtype: int64


In [138]:
numerical_cols = data.select_dtypes(include=['float64', 'int64']).columns.tolist()
categorical_cols = data.select_dtypes(include=['object', 'datetime']).columns.tolist()

print("Kolom Numerikal:", numerical_cols)
print("Kolom Kategorikal:", categorical_cols)

Kolom Numerikal: ['Rating', 'Reviews', 'Installs', 'Price', 'Cluster']
Kolom Kategorikal: ['App', 'Category', 'Size', 'Type', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [139]:
# Label encoder
label_encoder = LabelEncoder()
for col in categorical_cols:
    data[col] = label_encoder.fit_transform(data[col])

data.head()

App  Category  Rating   Reviews  Size    Installs  Type  Price  \
0  5805         0     4.1     159.0    47     10000.0     0    0.0   
1  2134         0     3.9     967.0    24    500000.0     0    0.0   
2  7321         0     4.7   87510.0   333   5000000.0     0    0.0   
3  6572         0     4.5  215644.0    87  50000000.0     0    0.0   
4  5860         0     4.3     967.0    56    100000.0     0    0.0   

   Content Rating  Genres  Last Updated  Current Ver  Android Ver  Cluster  
0               1       9          1086          107           14        0  
1               1      11          1094          952           14        1  
2               1       9          1291          424           14        0  
3               4       9          1237         2582           17        0  
4               1      10          1249          250           19        0

In [140]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8886 entries, 0 to 8885
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8886 non-null   int64  
 1   Category        8886 non-null   int64  
 2   Rating          8886 non-null   float64
 3   Reviews         8886 non-null   float64
 4   Size            8886 non-null   int64  
 5   Installs        8886 non-null   float64
 6   Type            8886 non-null   int64  
 7   Price           8886 non-null   float64
 8   Content Rating  8886 non-null   int64  
 9   Genres          8886 non-null   int64  
 10  Last Updated    8886 non-null   int64  
 11  Current Ver     8886 non-null   int64  
 12  Android Ver     8886 non-null   int64  
 13  Cluster         8886 non-null   int64  
dtypes: float64(4), int64(10)
memory usage: 972.0 KB


# **3. Data Splitting**

Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [141]:
scaler = StandardScaler()
data[numerical_cols] = scaler.fit_transform(data[numerical_cols])

X = data.drop(['Cluster'], axis=1)
y = data['Cluster']

# membagi dataset untuk dilatih 80% dan ditest 20%
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=42)

# Memastikan jumlah set
print(f"Jumlah data latih: {len(X_train)}")
print(f"Jumlah data uji: {len(X_test)}")

Jumlah data latih: 5331
Jumlah data uji: 3555


In [142]:
# Mengubah tipedata cluster menjadi integer
y_train = y_train.astype(int)
y_test = y_test.astype(int)

# Cek tipedata target
print(y_train.dtype)
print(y_test.dtype)

int64
int64


# **4. Membangun Model Klasifikasi**


## **a. Membangun Model Klasifikasi**

Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Pilih algoritma klasifikasi yang sesuai, seperti Logistic Regression, Decision Tree, Random Forest, atau K-Nearest Neighbors (KNN).
2. Latih model menggunakan data latih.

In [143]:
models = {
    'Decision Tree': DecisionTreeClassifier(),
    'Random Forest': RandomForestClassifier(),
    'Logistic Regression': LogisticRegression(),
    'K-Nearest Neighbors': KNeighborsClassifier(),
    'Support Vector Machine': SVC()
}

Tulis narasi atau penjelasan algoritma yang Anda gunakan.

## **b. Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Lakukan prediksi menggunakan data uji.
2. Hitung metrik evaluasi seperti Accuracy dan F1-Score (Opsional: Precision dan Recall).
3. Buat confusion matrix untuk melihat detail prediksi benar dan salah.

In [144]:
# Normalisasi data
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Melatih dan mengevaluasi setiap model
for model_name, classifier in models.items():
    try:
        # Melatih model dengan data latih
        if model_name == "Logistic Regression":
            classifier.set_params(max_iter=500, solver='saga')  # Meningkatkan iterasi & solver alternatif

        classifier.fit(X_train_scaled, y_train)  # Menggunakan data yang sudah dinormalisasi

        # Memprediksi hasil menggunakan data uji
        y_pred = classifier.predict(X_test_scaled)

        # Menghitung akurasi model
        model_accuracy = accuracy_score(y_test, y_pred)

        # Membuat laporan klasifikasi
        class_report = classification_report(y_test, y_pred)

        # Membuat matriks kebingungan (confusion matrix)
        confusion_mat = confusion_matrix(y_test, y_pred)

        # Menampilkan hasil evaluasi
        print(f"Model: {model_name}")
        print(f"Akurasi: {model_accuracy:.4f}")
        print("Laporan Klasifikasi:\n", class_report)
        print("Matriks Kebingungan:\n", confusion_mat)
        print("=" * 50)

    except Exception as e:
        print(f"Model {model_name} gagal dieksekusi. Error: {e}")


Model: Decision Tree
Akurasi: 0.9997
Laporan Klasifikasi:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00      2752
           1       1.00      1.00      1.00       755
           3       1.00      1.00      1.00        43
           5       1.00      1.00      1.00         5

    accuracy                           1.00      3555
   macro avg       1.00      1.00      1.00      3555
weighted avg       1.00      1.00      1.00      3555

Matriks Kebingungan:
 [[2751    1    0    0]
 [   0  755    0    0]
 [   0    0   43    0]
 [   0    0    0    5]]
Model: Random Forest
Akurasi: 0.9994
Laporan Klasifikasi:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00      2752
           1       1.00      1.00      1.00       755
           3       1.00      1.00      1.00        43
           5       1.00      1.00      1.00         5

    accuracy                           1.00      3555
   mac

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


Model: Logistic Regression
Akurasi: 0.9977
Laporan Klasifikasi:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00      2752
           1       1.00      1.00      1.00       755
           3       0.95      0.93      0.94        43
           5       1.00      1.00      1.00         5

    accuracy                           1.00      3555
   macro avg       0.99      0.98      0.98      3555
weighted avg       1.00      1.00      1.00      3555

Matriks Kebingungan:
 [[2747    3    2    0]
 [   0  755    0    0]
 [   3    0   40    0]
 [   0    0    0    5]]
Model: K-Nearest Neighbors
Akurasi: 0.9305
Laporan Klasifikasi:
               precision    recall  f1-score   support

           0       0.93      0.99      0.96      2752
           1       0.94      0.72      0.82       755
           3       1.00      0.88      0.94        43
           5       1.00      1.00      1.00         5

    accuracy                           0.93     

Tulis hasil evaluasi algoritma yang digunakan, jika Anda menggunakan 2 algoritma, maka bandingkan hasilnya.

1. Decision Tree dan Random Forest adalah model dengan akurasi tertinggi (≥99.94%).
Kedua model ini menunjukkan precision, recall, dan f1-score sempurna (1.00). Decision Tree sedikit lebih baik dari Random Forest, tetapi perbedaannya kecil. Potensi overfitting bisa menjadi masalah karena hasilnya terlalu sempurna.

2. Logistic Regression masih cukup kompetitif dengan akurasi 99.77%. Precision, recall, dan f1-score hampir sempurna. Namun, ada peringatan “max_iter was reached”, menunjukkan model mungkin belum sepenuhnya konvergen. Bisa diperbaiki dengan meningkatkan jumlah iterasi atau menggunakan solver yang lebih baik.

3. K-Nearest Neighbors (KNN) memiliki performa paling rendah. Akurasi 99.35% masih tinggi, tetapi lebih rendah dibanding model lain. Precision dan recall lebih rendah, terutama pada kelas minoritas, menunjukkan sensitivitas terhadap distribusi data.

4. Support Vector Machine (SVM) memiliki akurasi tinggi (99.10%). Performa cukup stabil dengan precision dan recall 99%. Namun, SVM biasanya lebih lambat dibanding model lain, terutama pada dataset besar.


## **c. Tuning Model Klasifikasi (Optional)**

Gunakan GridSearchCV, RandomizedSearchCV, atau metode lainnya untuk mencari kombinasi hyperparameter terbaik

In [153]:
# Decision Tree
param_grid_dt = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Membuat objek GridSearchCV
grid_search_dt = GridSearchCV(
    estimator=DecisionTreeClassifier(),
    param_grid=param_grid_dt,
    cv=5,
    scoring='accuracy',
    n_jobs=-1
)

# Melatih GridSearchCV
grid_search_dt.fit(X_train_scaled, y_train)

# Menampilkan hasil terbaik
print("Decision Tree")
print(f"Best Parameters: {grid_search_dt.best_params_}")
print(f"Best Score: {grid_search_dt.best_score_:.4f}")
print("=" * 50)

Decision Tree
Best Parameters: {'criterion': 'entropy', 'max_depth': 20, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best Score: 0.9996


In [154]:
# Random Forest
param_grid_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

grid_search_rf = GridSearchCV(
    estimator=RandomForestClassifier(),
    param_grid=param_grid_rf,
    cv=5,
    scoring='accuracy',
    n_jobs=-1
)

grid_search_rf.fit(X_train_scaled, y_train)

print("Random Forest")
print(f"Best Parameters: {grid_search_rf.best_params_}")
print(f"Best Score: {grid_search_rf.best_score_:.4f}")
print("=" * 50)

Random Forest
Best Parameters: {'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 5, 'n_estimators': 200}
Best Score: 0.9994


In [155]:
# Logistic Regression
param_grid_lr = {
    'C': [0.01, 0.1, 1, 10, 100],
    'solver': ['lbfgs', 'saga']
}

grid_search_lr = GridSearchCV(
    estimator=LogisticRegression(max_iter=500),
    param_grid=param_grid_lr,
    cv=5,
    scoring='accuracy',
    n_jobs=-1
)

grid_search_lr.fit(X_train_scaled, y_train)

print("Logistic Regression")
print(f"Best Parameters: {grid_search_lr.best_params_}")
print(f"Best Score: {grid_search_lr.best_score_:.4f}")
print("=" * 50)

Logistic Regression
Best Parameters: {'C': 10, 'solver': 'lbfgs'}
Best Score: 0.9996


In [159]:
# knn
param_grid_knn = {
    'n_neighbors': [3, 5, 7, 9],
    'weights': ['uniform', 'distance'],
    'metric': ['euclidean', 'manhattan']
}

grid_search_knn = GridSearchCV(
    estimator=KNeighborsClassifier(),
    param_grid=param_grid_knn,
    cv=5,
    scoring='accuracy',
    n_jobs=-1
)

grid_search_knn.fit(X_train_scaled, y_train)

print("K-Nearest Neighbors")
print(f"Best Parameters: {grid_search_knn.best_params_}")
print(f"Best Score: {grid_search_knn.best_score_:.4f}")
print("=" * 50)

K-Nearest Neighbors
Best Parameters: {'metric': 'manhattan', 'n_neighbors': 7, 'weights': 'distance'}
Best Score: 0.9379


In [157]:
# SVM
param_grid_svm = {
    'C': [0.1, 1, 10, 100],
    'kernel': ['linear', 'rbf', 'poly'],
    'gamma': ['scale', 'auto']
}

grid_search_svm = GridSearchCV(
    estimator=SVC(),
    param_grid=param_grid_svm,
    cv=5,
    scoring='accuracy',
    n_jobs=-1
)

grid_search_svm.fit(X_train_scaled, y_train)

print("SVM")
print(f"Best Parameters: {grid_search_svm.best_params_}")
print(f"Best Score: {grid_search_svm.best_score_:.4f}")
print("=" * 50)

SVM
Best Parameters: {'C': 10, 'gamma': 'scale', 'kernel': 'linear'}
Best Score: 0.9998


## **d. Evaluasi Model Klasifikasi setelah Tuning (Optional)**

Berikut adalah rekomendasi tahapannya.
1. Gunakan model dengan hyperparameter terbaik.
2. Hitung ulang metrik evaluasi untuk melihat apakah ada peningkatan performa.

1. Saya menggunakan GridSearchCV, karena dataset nya tidak terlalu besar.
2. Setelah dihitung ulang metrik evaluasinya, dapat disimpulkan bahwa:
  - Logistic Regression mengalami peningkatan akurasi signifikan (+0.19%), menunjukkan tuning hyperparameter berhasil meningkatkan performanya.
  - SVM meningkat cukup signifikan (+0.88%), menunjukkan bahwa pemilihan kernel linear dengan C=10 lebih optimal.
  - Decision Tree mengalami sedikit penurunan (-0.01%), tetapi masih sangat akurat (99.96%).
  - Random Forest tetap stabil di 99.94%, menunjukkan bahwa hyperparameter tuning tidak selalu menghasilkan perubahan besar pada model yang sudah optimal.
  - K-Nearest Neighbors mengalami penurunan akurasi (-5.56%), kemungkinan karena pemilihan parameter n_neighbors=7 dan weights=distance kurang optimal untuk dataset ini.

## **e. Analisis Hasil Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Bandingkan hasil evaluasi sebelum dan setelah tuning (jika dilakukan).
2. Identifikasi kelemahan model, seperti:
  - Precision atau Recall rendah untuk kelas tertentu.
  - Apakah model mengalami overfitting atau underfitting?
3. Berikan rekomendasi tindakan lanjutan, seperti mengumpulkan data tambahan atau mencoba algoritma lain jika hasil belum memuaskan.

1. Insight:
  - SVM mengalami peningkatan terbesar (+0.88%), menjadi model dengan performa terbaik setelah tuning.
  - Logistic Regression juga meningkat (+0.19%), menunjukkan tuning efektif meningkatkan kestabilan model.
  - Decision Tree & Random Forest tidak mengalami perubahan signifikan, menunjukkan hyperparameter default sudah cukup optimal.
  - KNN mengalami penurunan drastis (-5.56%), kemungkinan karena tuning hyperparameter tidak cocok dengan distribusi data.

2. - Decision Tree & Random Forest berpotensi overfitting karena akurasi hampir sempurna pada dataset ini.
  - SVM & Logistic Regression memiliki keseimbangan yang baik, menjadikannya pilihan terbaik untuk generalisasi.
  - KNN mengalami underfitting setelah tuning, sehingga perlu ditinjau ulang parameter n_neighbors dan metrik jaraknya.

3. Gunakan SVM atau Logistic Regression sebagai model utama karena memiliki performa tinggi dan generalisasi yang baik.
  